File: ui/src/types/conversations.ts
/**
 * TypeScript types for conversation API responses.
 */

// Type aliases matching backend Literal types
export type ConversationStatus = 'active' | 'archived' | 'deleted'
export type MessageRole = 'user' | 'assistant' | 'system' | 'tool'
export type MessageStatus = 'pending' | 'streaming' | 'complete' | 'failed'
export type FeedbackRating = 'positive' | 'negative'

export interface ConversationResponse {
  id: string
  tenant_id: string
  user_id: string
  scope_type: string
  scope_id: string | null
  module: string | null
  title: string | null
  status: ConversationStatus
  message_count: number
  total_input_tokens: number
  total_output_tokens: number
  total_cost_micros: number
  session_id: string | null
  run_id: string | null
  created_at: string
  updated_at: string
  last_message_at: string | null
}

export interface ConversationListResponse {
  items: ConversationResponse[]
  total: number
}

export interface MessageResponse {
  id: string
  conversation_id: string
  role: MessageRole
  content: string | null
  parts: Record<string, unknown> | null
  input_tokens: number | null
  output_tokens: number | null
  cost_micros: number | null
  latency_ms: number | null
  model_id: string | null
  status: MessageStatus
  parent_message_id: string | null
  sequence_number: number
  created_at: string
}

export interface MessageListResponse {
  items: MessageResponse[]
  total: number
}

export interface FeedbackResponse {
  id: string
  message_id: string
  user_id: string
  rating: FeedbackRating
  content: string | null
  created_at: string
}

export interface BranchResponse {
  conversation_id: string
  branch_point_message_id: string
  new_sequence_number: number
}

export interface SiblingResponse {
  items: MessageResponse[]
  total: number
  current_index: number
}

File: src/intelli/schemas/conversations.py
"""Pydantic schemas for Conversations and Messages."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# Type aliases for status fields
ConversationStatus = Literal["active", "archived", "deleted"]
MessageRole = Literal["user", "assistant", "system", "tool"]
MessageStatus = Literal["pending", "streaming", "complete", "failed"]
FeedbackRating = Literal["positive", "negative"]


# ---------------------------------------------------------------------------
# Conversation schemas
# ---------------------------------------------------------------------------


class ConversationCreate(BaseModel):
    """Request to create a new conversation."""

    scope_type: str = Field(default="tenant", description="Scope level binding")
    scope_id: UUID | None = Field(None, description="Scope entity ID (null for tenant-level)")
    module: str | None = Field(None, description="Module context: research, analysis, etc.")
    title: str | None = Field(None, description="Conversation title (auto-generated if absent)")
    session_id: UUID | None = Field(None, description="Session to bind this conversation to")
    config_snapshot: dict = Field(default_factory=dict, description="Resolved config snapshot")

    model_config = ConfigDict(extra="forbid")


class ConversationUpdate(BaseModel):
    """Request to update a conversation."""

    title: str | None = None
    status: ConversationStatus | None = Field(None, description="Conversation status")

    model_config = ConfigDict(extra="forbid")


class ConversationResponse(BaseModel):
    """Conversation in API responses."""

    id: UUID
    tenant_id: UUID
    user_id: UUID
    scope_type: str
    scope_id: UUID | None
    module: str | None
    title: str | None
    status: ConversationStatus
    message_count: int
    total_input_tokens: int
    total_output_tokens: int
    total_cost_micros: int
    session_id: UUID | None
    run_id: UUID | None
    created_at: datetime
    updated_at: datetime
    last_message_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class ConversationListResponse(BaseModel):
    """Paginated list of conversations."""

    items: list[ConversationResponse]
    total: int


# ---------------------------------------------------------------------------
# Message schemas
# ---------------------------------------------------------------------------


class MessageResponse(BaseModel):
    """Message in API responses."""

    id: UUID
    conversation_id: UUID
    role: MessageRole
    content: str | None
    parts: dict | None
    input_tokens: int | None
    output_tokens: int | None
    cost_micros: int | None
    latency_ms: int | None
    model_id: str | None
    status: MessageStatus
    parent_message_id: UUID | None
    sequence_number: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class MessageListResponse(BaseModel):
    """Paginated list of messages."""

    items: list[MessageResponse]
    total: int


# ---------------------------------------------------------------------------
# Feedback schemas
# ---------------------------------------------------------------------------


class FeedbackCreate(BaseModel):
    """Request to add feedback to a message."""

    rating: FeedbackRating = Field(..., description="Feedback rating")
    content: str | None = Field(None, description="Optional explanation")

    model_config = ConfigDict(extra="forbid")


class FeedbackResponse(BaseModel):
    """Feedback in API responses."""

    id: UUID
    message_id: UUID
    user_id: UUID
    rating: FeedbackRating
    content: str | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Branch schemas
# ---------------------------------------------------------------------------


class BranchCreate(BaseModel):
    """Request to branch from a message."""

    message_id: UUID = Field(..., description="Message to branch from")

    model_config = ConfigDict(extra="forbid")


class BranchResponse(BaseModel):
    """Response after creating a branch."""

    conversation_id: UUID
    branch_point_message_id: UUID
    new_sequence_number: int


class SiblingResponse(BaseModel):
    """Response containing sibling messages."""

    items: list[MessageResponse]
    total: int
    current_index: int
